src/command/restore/restore.c

```diff
@@ -0,0 +1,1744 @@
+
+        // Skip the tablespace_map file when present so PostgreSQL does not rewrite links in pg_tblspc. The tablespace links will be
+        // created after paths are cleaned.
+        if (manifestFileFindDefault(manifest, STRDEF(MANIFEST_TARGET_PGDATA "/" PG_FILE_TABLESPACEMAP), NULL) != NULL &&
+            manifestData(manifest)->pgVersion >= PG_VERSION_TABLESPACE_MAP)
+        {
+            LOG_DETAIL("skip '"PG_FILE_TABLESPACEMAP"' -- tablespace links will be created based on mappings");
+            manifestFileRemove(manifest, STRDEF(MANIFEST_TARGET_PGDATA "/" PG_FILE_TABLESPACEMAP));
+        }
+
+        // Step 2: Clean target directories
+        // -------------------------------------------------------------------------------------------------------------------------
+        for (unsigned int targetIdx = 0; targetIdx < manifestTargetTotal(manifest); targetIdx++)
+        {
+            RestoreCleanCallbackData *cleanData = &cleanDataList[targetIdx];
+
+            // Only clean if the target exists
+            if (cleanData->exists)
+            {
+                // Don't clean file links.  It doesn't matter whether the file exists or not since we know it is in the manifest.
+                if (cleanData->target->file == NULL)
+                {
+                    // Only log when doing a delta restore because otherwise the targets should be empty.  We'll still run the clean
+                    // to fix permissions/ownership on the target paths.
+                    if (delta)
+                        LOG_INFO("remove invalid files/links/paths from '%s'", strPtr(cleanData->targetPath));
+
+                    // Clean the target
+                    storageInfoListP(
+                        storageLocalWrite(), cleanData->targetPath, restoreCleanInfoListCallback, cleanData, .errorOnMissing = true,
+                        .sortOrder = sortOrderAsc);
+                }
+            }
+            // If the target does not exist we'll attempt to create it
+            else
+            {
+                const ManifestPath *path = NULL;
+
+                // There is no path information for a file link so we'll need to use the data directory
+                if (cleanData->target->file != NULL)
+                {
+                    path = manifestPathFind(manifest, MANIFEST_TARGET_PGDATA_STR);
+                }
+                // Else grab the info for the path that matches the link name
+                else
+                    path = manifestPathFind(manifest, cleanData->target->name);
+
+                storagePathCreateP(storageLocalWrite(), cleanData->targetPath, .mode = path->mode);
+                restoreCleanOwnership(cleanData->targetPath, path->user, path->group, userId(), groupId(), true);
+            }
+        }
+
+        // Step 3: Create missing paths and path links
+        // -------------------------------------------------------------------------------------------------------------------------
```

Comment:
Because we have now built out enough structure to always be able to use paths relative to pg data.  This is safer because we'll get an error if we accidentally write outside of pg data.

We'd like to do this in step 2 as well, but there's a chicken and egg problem there (we need to cleanup before we can create the links, but we need the links to address pg data as one tree).  It might be resolvable but it was taking too much time so seemed better to leave if for later if at all.

Reply:
Ok